src/js/panels.js

```diff
@@ -1101,10 +1677,66 @@ https://raw.githubusercontent.com/GPII/first-discovery/master/LICENSE.txt
             savePrefsToServer: {
                 funcName: "gpii.firstDiscovery.panel.token.savePrefsToServer",
                 args: ["{that}", "{fluid.prefs.prefsEditor}.model.preferences", "{arguments}.0"]
+            },
+            savePrefsToUsb: {
+                funcName: "gpii.firstDiscovery.panel.token.savePrefsToUsb",
+                args: ["{that}", "{arguments}.0"]
+
             }
         }
     });
 
+    gpii.firstDiscovery.panel.token.savePrefsToUsb = function (that, data) {
+
+        /*
+         if not chrome:
+         USB is not possible, so just fire success
+
+         if chrome:
+
+         if not extension:
+         USB is not possible so just fire success
+
+         if extension:
+         USB is possible so msg the extension
+         */
+
+        // If the plugin hasn't responded in 2 seconds with the version, it is a user without a
+        // plugin installed, so show them the token.
+        var successTimeoutId = setTimeout(function () {
+            that.events.onSuccess.fire(data);
+        }, 2000);
+
+        window.chrome.runtime.sendMessage(
+            that.options.chromeExtensionId,
+            {"message": {"message_type": "request_version"}},
+            function onVersionCallback(version) {
+                // If we got a version, the plugin is installed, so clear the timeout and let the plugin write to usb
+                if (version) {
+                    clearTimeout(successTimeoutId);
+                    var message = {
+                        message: {
+                            message_type: "write_usb",
+                            message_body: {
+                                userToken: data.userToken,
+                                preferences: data.preferences
+                            }
+                        }
+                    };
+                    window.chrome.runtime.sendMessage(
```

Comment:
Break out all of this deeply nested callback material into an event-driven architecture using suitable global functions